File: server/src/services/githubService.js
/**
 * GitHub API Service
 * Handles all GitHub API interactions using Octokit
 */

const { Octokit } = require('octokit');

class GitHubService {
    constructor(token = null) {
        // Use provided token or fall back to PAT from env
        const authToken = token || process.env.GITHUB_PAT;

        if (!authToken) {
            throw new Error('GitHub token not provided');
        }

        this.octokit = new Octokit({ auth: authToken });
    }

    /**
     * Get authenticated user's profile
     */
    async getUser() {
        const { data } = await this.octokit.rest.users.getAuthenticated();
        return {
            username: data.login,
            name: data.name,
            avatarUrl: data.avatar_url,
            bio: data.bio,
            publicRepos: data.public_repos,
            followers: data.followers,
            following: data.following,
        };
    }

    /**
     * Get user's repositories
     * @param {string} username - GitHub username
     * @param {number} perPage - Number of repos to fetch
     */
    async getRepos(username, perPage = 10) {
        const { data } = await this.octokit.rest.repos.listForUser({
            username,
            sort: 'updated',
            per_page: perPage,
        });

        return data.map((repo) => ({
            name: repo.name,
            fullName: repo.full_name,
            description: repo.description,
            language: repo.language,
            stars: repo.stargazers_count,
            forks: repo.forks_count,
            updatedAt: repo.updated_at,
            url: repo.html_url,
            isPrivate: repo.private,
        }));
    }

    /**
     * Get recent commits across user's repos
     * @param {string} username - GitHub username
     * @param {number} days - Number of days to look back
     */
    async getRecentCommits(username, days = 7) {
        const since = new Date();
        since.setDate(since.getDate() - days);

        try {
            // Get user's recent events (includes push events with commits)
            const { data: events } = await this.octokit.rest.activity.listPublicEventsForUser({
                username,
                per_page: 100,
            });

            const commits = [];

            for (const event of events) {
                if (event.type === 'PushEvent' && event.payload.commits) {
                    for (const commit of event.payload.commits) {
                        commits.push({
                            sha: commit.sha.substring(0, 7),
                            message: commit.message.split('\n')[0], // First line only
                            repo: event.repo.name,
                            date: event.created_at,
                            url: `https://github.com/${event.repo.name}/commit/${commit.sha}`,
                        });
                    }
                }
            }

            // Filter by date and limit
            const filteredCommits = commits.filter(
                (c) => new Date(c.date) >= since
            );

            return filteredCommits.slice(0, 50); // Limit to 50 commits
        } catch (error) {
            console.error('Error fetching commits:', error.message);
            return [];
        }
    }

    /**
     * Get user's contribution activity summary
     * @param {string} username - GitHub username
     */
    async getActivitySummary(username) {
        try {
            const { data: events } = await this.octokit.rest.activity.listPublicEventsForUser({
                username,
                per_page: 100,
            });

            const today = new Date();
            today.setHours(0, 0, 0, 0);

            const summary = {
                totalEvents: events.length,
                todayEvents: 0,
                pushEvents: 0,
                prEvents: 0,
                issueEvents: 0,
                reposWorkedOn: new Set(),
                languagesUsed: new Set(),
            };

            for (const event of events) {
                const eventDate = new Date(event.created_at);
                eventDate.setHours(0, 0, 0, 0);

                if (eventDate.getTime() === today.getTime()) {
                    summary.todayEvents++;
                }

                summary.reposWorkedOn.add(event.repo.name);

                switch (event.type) {
                    case 'PushEvent':
                        summary.pushEvents++;
                        break;
                    case 'PullRequestEvent':
                        summary.prEvents++;
                        break;
                    case 'IssuesEvent':
                        summary.issueEvents++;
                        break;
                }
            }

            return {
                ...summary,
                reposWorkedOn: Array.from(summary.reposWorkedOn),
                reposCount: summary.reposWorkedOn.size,
            };
        } catch (error) {
            console.error('Error fetching activity summary:', error.message);
            throw error;
        }
    }

    /**
     * Get languages used in user's repos
     * @param {string} username - GitHub username
     */
    async getLanguages(username) {
        const repos = await this.getRepos(username, 20);
        const languages = {};

        for (const repo of repos) {
            if (repo.language) {
                languages[repo.language] = (languages[repo.language] || 0) + 1;
            }
        }

        // Sort by count and return as array
        return Object.entries(languages)
            .sort((a, b) => b[1] - a[1])
            .map(([name, count]) => ({ name, count }));
    }
}

module.exports = GitHubService;

File: flutter_app/lib/services/github_service.dart
import '../config/api_config.dart';
import 'api_service.dart';

/// Service for GitHub-related operations
class GitHubService {
  final ApiService _api = ApiService();

  /// Get GitHub profile
  Future<GitHubProfile> getProfile() async {
    try {
      final response = await _api.get(ApiEndpoints.githubProfile);
      return GitHubProfile.fromJson(response.data);
    } catch (e) {
      print('Error fetching GitHub profile: $e');
      rethrow;
    }
  }

  /// Get user's repositories
  Future<List<GitHubRepo>> getRepositories() async {
    try {
      final response = await _api.get(ApiEndpoints.githubRepos);
      final List<dynamic> data = response.data['repos'] ?? response.data ?? [];
      return data.map((e) => GitHubRepo.fromJson(e)).toList();
    } catch (e) {
      print('Error fetching repositories: $e');
      rethrow;
    }
  }

  /// Get programming languages used
  Future<Map<String, int>> getLanguages() async {
    try {
      final response = await _api.get(ApiEndpoints.githubLanguages);
      return Map<String, int>.from(response.data);
    } catch (e) {
      print('Error fetching languages: $e');
      rethrow;
    }
  }

  /// Analyze a specific repository
  Future<RepoAnalysis> analyzeRepo(String owner, String repo) async {
    try {
      final response = await _api.get(ApiEndpoints.githubRepo(owner, repo));
      return RepoAnalysis.fromJson(response.data);
    } catch (e) {
      print('Error analyzing repo $owner/$repo: $e');
      rethrow;
    }
  }

  /// Get recent GitHub activity
  Future<List<GitHubActivity>> getActivity() async {
    try {
      final response = await _api.get(ApiEndpoints.githubActivity);
      final List<dynamic> data = response.data['activity'] ?? response.data ?? [];
      return data.map((e) => GitHubActivity.fromJson(e)).toList();
    } catch (e) {
      print('Error fetching activity: $e');
      rethrow;
    }
  }

  /// Get commit history
  Future<List<GitHubCommit>> getCommits({int limit = 20}) async {
    try {
      final response = await _api.get(
        ApiEndpoints.githubCommits,
        queryParameters: {'limit': limit},
      );
      final List<dynamic> data = response.data['commits'] ?? response.data ?? [];
      return data.map((e) => GitHubCommit.fromJson(e)).toList();
    } catch (e) {
      print('Error fetching commits: $e');
      rethrow;
    }
  }
}

// ==================== MODELS ====================

class GitHubProfile {
  final String login;
  final String name;
  final String? avatarUrl;
  final String? bio;
  final int publicRepos;
  final int followers;
  final int following;

  const GitHubProfile({
    required this.login,
    required this.name,
    this.avatarUrl,
    this.bio,
    this.publicRepos = 0,
    this.followers = 0,
    this.following = 0,
  });

  factory GitHubProfile.fromJson(Map<String, dynamic> json) {
    return GitHubProfile(
      login: json['login'] ?? '',
      name: json['name'] ?? json['login'] ?? '',
      avatarUrl: json['avatar_url'] ?? json['avatarUrl'],
      bio: json['bio'],
      publicRepos: json['public_repos'] ?? json['publicRepos'] ?? 0,
      followers: json['followers'] ?? 0,
      following: json['following'] ?? 0,
    );
  }
}

class GitHubRepo {
  final String name;
  final String fullName;
  final String? description;
  final String htmlUrl;
  final String? language;
  final int stars;
  final int forks;
  final bool isPrivate;
  final DateTime? updatedAt;

  const GitHubRepo({
    required this.name,
    required this.fullName,
    this.description,
    required this.htmlUrl,
    this.language,
    this.stars = 0,
    this.forks = 0,
    this.isPrivate = false,
    this.updatedAt,
  });

  factory GitHubRepo.fromJson(Map<String, dynamic> json) {
    return GitHubRepo(
      name: json['name'] ?? '',
      fullName: json['full_name'] ?? json['fullName'] ?? '',
      description: json['description'],
      htmlUrl: json['html_url'] ?? json['htmlUrl'] ?? '',
      language: json['language'],
      stars: json['stargazers_count'] ?? json['stars'] ?? 0,
      forks: json['forks_count'] ?? json['forks'] ?? 0,
      isPrivate: json['private'] ?? json['isPrivate'] ?? false,
      updatedAt: json['updated_at'] != null 
          ? DateTime.parse(json['updated_at']) 
          : null,
    );
  }
}

class RepoAnalysis {
  final String name;
  final List<String> languages;
  final int commits;
  final int contributors;
  final Map<String, int> commitPatterns;
  final List<String> technologies;

  const RepoAnalysis({
    required this.name,
    this.languages = const [],
    this.commits = 0,
    this.contributors = 0,
    this.commitPatterns = const {},
    this.technologies = const [],
  });

  factory RepoAnalysis.fromJson(Map<String, dynamic> json) {
    return RepoAnalysis(
      name: json['name'] ?? '',
      languages: json['languages'] != null 
          ? List<String>.from(json['languages']) 
          : [],
      commits: json['commits'] ?? 0,
      contributors: json['contributors'] ?? 0,
      commitPatterns: json['commitPatterns'] != null 
          ? Map<String, int>.from(json['commitPatterns']) 
          : {},
      technologies: json['technologies'] != null 
          ? List<String>.from(json['technologies']) 
          : [],
    );
  }
}

class GitHubActivity {
  final String type;
  final String repo;
  final String? message;
  final DateTime createdAt;

  const GitHubActivity({
    required this.type,
    required this.repo,
    this.message,
    required this.createdAt,
  });

  factory GitHubActivity.fromJson(Map<String, dynamic> json) {
    return GitHubActivity(
      type: json['type'] ?? '',
      repo: json['repo'] ?? '',
      message: json['message'],
      createdAt: json['created_at'] != null 
          ? DateTime.parse(json['created_at']) 
          : DateTime.now(),
    );
  }
}

class GitHubCommit {
  final String sha;
  final String message;
  final String repo;
  final DateTime committedAt;

  const GitHubCommit({
    required this.sha,
    required this.message,
    required this.repo,
    required this.committedAt,
  });

  factory GitHubCommit.fromJson(Map<String, dynamic> json) {
    return GitHubCommit(
      sha: json['sha'] ?? '',
      message: json['message'] ?? '',
      repo: json['repo'] ?? '',
      committedAt: json['committed_at'] != null 
          ? DateTime.parse(json['committed_at']) 
          : DateTime.now(),
    );
  }
}
